Fix Alchemy.user_class_name constant conflict

mickenorlen opened the pull request:
## What is this pull request for?
After fresh installation in rails 6, after creating the index page, I got this error when trying to load page tree:
```
[active_model_serializers] Rendered ActiveModel::Serializer::Null with Alchemy::PageTreeSerializer (26.56ms)

NoMethodError undefined method `arel_table' for Alchemy::Admin:Module in /app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/table_metadata.rb:51:in `associated_table'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/predicate_builder.rb:77:in `block in expand_from_hash'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/predicate_builder.rb:68:in `each'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/predicate_builder.rb:68:in `flat_map'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/predicate_builder.rb:68:in `expand_from_hash'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/predicate_builder.rb:21:in `build_from_hash'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/where_clause_factory.rb:19:in `build'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/query_methods.rb:656:in `where!'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/relation/query_methods.rb:649:in `where'
/app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/querying.rb:21:in `where'
/app/_docker/data/bundle/ruby/2.7.0/gems/alchemy_cms-4.4.1/app/models/alchemy/folded_page.rb:20:in `folded_for_user'
/app/_docker/data/bundle/ruby/2.7.0/gems/alchemy_cms-4.4.1/app/serializers/alchemy/page_tree_serializer.rb:15:in `pages'
/app/_docker/data/bundle/ruby/2.7.0/gems/alchemy_cms-4.4.1/app/serializers/alchemy/base_serializer.rb:26:in `read_attribute_for_serialization'

```

Then I found this which seems to be a working solution: 
https://gitlab.com/gitlab-org/gitlab-foss/-/merge_requests/18329
https://gitlab.com/gitlab-org/gitlab-foss/-/merge_requests/18329/diffs

### Notable change
Only change made is that I replaced all instances of:
`Alchemy.user_class_name`
with:
`"::#{Alchemy.user_class_name}"`


I just started to test alchemy out so I can't tell if this has some other consequences as of yet. Also I'm not sure the `stamper_class_name` should be replaced in the same manner.

tvdeyen commented:
Thanks. I think that makes sense to avoid constant name conflicts. 

But could you please make the change in the 'lib/alchemy/auth_accessors.rb` file instead? The `user_class_name` method is already a String and it will reduce the number of files that needs to be changed down to one.

mickenorlen commented:
Yeah, that makes more sense. 

I made getter/setter methods for the user_class_name variable and removed it from mattr_accessor. Just tell me if there's something you want to change.

mickenorlen commented:
Btw, i think this might need a bit more testing.. I'm not sure if its related but now i'm seeing this when I'm trying to fold pages:

```
ActiveRecord::InverseOfAssociationNotFoundError Could not find the inverse association for user (:folded_pages in ::Admin) in /app/_docker/data/bundle/ruby/2.7.0/gems/activerecord-6.0.2.1/lib/active_record/reflection.rb:240:in `check_validity_of_inverse!'
```

tvdeyen commented on lib/alchemy/auth_accessors.rb:
```diff
@@ -73,6 +72,16 @@ module Alchemy
+
+  # Prefix with :: when getting to avoid constant name conflicts
+  def self.user_class_name
+    "::#{@@user_class_name}"
```
I think it makes sense to `.freeze` the string here explicitly

mickenorlen replied:
The hound said no freeze needed: https://github.com/AlchemyCMS/alchemy_cms/pull/1724#discussion_r382997756

tvdeyen replied:
Yeah. Forgot that we already freeze all strings. Please ignore my comment

tvdeyen commented:
Unfortunately the Github CI is having some issues. I will look into this soon. Meanwhile please run tests locally (they are very fast) and add a test for the namespace fix. Thanks again for your contribution

tvdeyen commented:
For the changelog (that gets generated from PR titles) it would be better to rename this PR

mickenorlen commented:
Ah, missed reading through the contribution notes. Is it possible to run Hound locally also?

Added a test for getting with prefix `::` and moved the old `user_class_name.is_a?(String)` test from `user_class` to `user_class_name` method.

Still have 2 failures. Unrelated?
```
Failures:

  1) Page request caching when caching is enabled in app but page should not be cached does not set etag header
     Failure/Error: expect(response.headers).to_not have_key('ETag')
       expected #has_key?("ETag") to return false, got true
     # ./spec/requests/alchemy/page_request_caching_spec.rb:103:in `block (4 levels) in <top (required)>'

  2) Page request caching when caching is enabled in app but a flash message is present does not set etag header
     Failure/Error: expect(response.headers).to_not have_key('ETag')
       expected #has_key?("ETag") to return false, got true
     # ./spec/requests/alchemy/page_request_caching_spec.rb:127:in `block (4 levels) in <top (required)>'
```

tvdeyen commented:
The two failing specs are unrelated.

tvdeyen commented:
@mickenorlen would you please rebase with latest `master` to fix the CI builds and squash the commits into one? Thanks
